### ISSUE TITLE
@copyOf should take a position on attributes

### ISSUE THREAD

**Issue by hcayless** [TEIC/TEI#2146]
The documentation for [`@copyOf`](https://www.tei-c.org/release/doc/tei-p5-doc/en/html/ref-att.global.linking.html#tei_att.copyOf) leaves it as an exercise for the reader whether or not attributes on the target element should be copied, noting only that
> Any content of the current element should be ignored. Its true content is that of the element being pointed at.

Logically, it would seem to me that if one element is a copy of another, it should have the attributes of the target as well as the content. `@xml:id` is the obvious exception to this, as an identical ID would raise a validation error. For example, if the target has an `@xml:lang`, presumably the copy should as well. Must this be repeated by the encoder, or can they assume `@copyOf` implies that the language setting applies to the copy?

The Guidelines (https://www.tei-c.org/release/doc/tei-p5-doc/en/html/SA.html#SAIE) do give an example where the copy has an additional `@rend` attribute. So there seems to have been some thought that attributes on the copy might differ.

I can think of the following solutions:

1. We say that the Guidelines deliberately leave to implementers the question of what to do with attributes.
2. All attributes on the target element are to be copied except `@xml:id` and any attributes on the pointing element are to be ignored, as with content.
3. Attributes from the target element are to be merged with those on the pointing element, so that the latter keeps any attributes it possesses and gains new ones from the target.
4. Attributes are not copied from the target element, so any desired attributes must be duplicated on the pointing element.

**Comment by bansp** [comment]
Option 3. could be beautiful. A little foray into a unification-based world.
Though I think it would then be good to mention that in the case of duplication, the local version (on the pointing element) wins.

**Comment by martindholmes** [comment]
Agreed on #3, which is more like an inherit + override system. But it does mean we're treating attributes differently from content, which is always ignored.

**Comment by hcayless** [comment]
I'm sort of neutral, except for really disliking option 4. The only use case I have for `@copyOf` is recording transpositions in variant readings (e.g. editor X put line 3 after line 10). In that case I don't duplicate the line, but mark the one in the variant reading as a copy. My preference would be for attributes on the target to carry over to the copy (except `@xml:id`).

**Comment by lb42** [comment]
If the content of the pointing element is to be ignored/suppressed, surely its attributes (if any) should be as well? In which case there's no problem in saying that all attributes of the targetted element are copied, along with its content.

**Comment by martindholmes** [comment]
@lb42 Except for `@copy-of`, presumably.

**Comment by lb42** [comment]
What? if there is a @copyOf on the targetted element, I think it's turtles all the way down!

**Comment by martindholmes** [comment]
There are implications for processing, though. If an `@copy-of` points to an element with another `@copy-of`, then two processing passes will probably be needed, or some recursion.

**Comment by lb42** [comment]
Yes. So better make sure your @copyOf processing is robust (it could for example refuse to handle a subsequent @copyOf) . But not a good reason to special case this (rather remote) possibility.

**Comment by martindholmes** [comment]
As far as I can see, the Stylesheets have no processing at all for `@copy-of`. Presumably something along these lines would be what we want:

~~~~~
<xsl:template match="*[@copy-of]" priority="1" mode="#all">
  <xsl:apply-templates select="xyz:resolvePointer(@copy-of)" mode="#current"/>
</xsl:template/>
~~~~~

with the tricky bit being the pointer resolution. That would handle any recursion, I think.

**Comment by hennyu** [comment]
@joeytakeda @torstenroeder and me suggest to go for option 2. proposed by @hcayless above, which was: "All attributes on the target element are to be copied except @xml:id and any attributes on the pointing element are to be ignored, as with content." We believe that this is clearer than mixing attributes from the copied and the target element. See the following example:

`
<date when="1997" rend="blue" xml:id="d1">1997</date>
<date when-iso="2017" copyOf="#d1" rend="brown">2017</date>
`

The option we see as the clearest one is:

`<date when="1997" rend="blue" sameAs="#d1">1997</date>`

**Comment by joeytakeda** [comment]
However, it looks like option 2, while the preferred one, might cause some problems with other examples. In the Trees chapter (which is referred to in the section defining copyOf), it seems problematic -- see https://www.tei-c.org/release/doc/tei-p5-doc/en/html/GD.html#GDAT

**Comment by ebeshero** [comment]
Council F2F Vancouver decision: `@copyOf` means this: Indicate only the contents of the element are to be copied  and not the attributes (which aligns with the examples we provide in the Guidelines). @torstenroeder  should update the prose to make this clear.
Sidenote: this conflicts with the meaning of `<xsl:copy-of>` which does explicitly copies over attributes (the entire node), but it is different for good reason, because we wouldn't want the original attributes like `@xml:id` to be copied over into a new context.

**Comment by ebeshero** [comment]
@joeytakeda notes that we should include a comment about using XInclude for other kinds of copying.

**Comment by ebeshero** [comment]
See https://github.com/TEIC/TEI/issues/2951 for a fix required as a result of this .

**Comment by hcayless** [comment]
I think that is a bad decision, fwiw. If you're going to insist on it, may I propose that the desired behavior be specified in the header. This is going to break other stuff heretofore valid.

**Comment by ebeshero** [comment]
@hcayless We took a lot of time to review the Guidelines prose and examples and found them consistent with this decision. We don't think anything will break as a result. In the Guidelines processing, nothing is going to break. People might have processing updates they might want to make, but we don't think there are serious problems.

**Comment by hcayless** [comment]
In the absence of a decision on this, people, including me, have essentially gone with option 1. So you've retroactively broken stuff based on, as far as I can tell, a chapter no one uses. You're going to do what you want, obviously, but I'd ask for a logical approach rather than a blind obedience to history.

**Comment by ebeshero** [comment]
Blind obedience to history? Hardly, because we did not agree with the comments on the ticket after we carefully unpacked and reviewed the examples in the Guidelines. We had a long discussion and a lot of debate about the examples and explanation given, and I think it's seriously unjust to say that this Council is unthinkingly obedient. We found the Guidelines consistent and indeed reasonable for not automatically copying attributes, and it took some care to work out the expected behavior of a placeholder element that indicates it stores a copy (not the same) of the contents of the node designated. We also found errors in a couple of examples related to this. It took some time to clear up that `@copyOf` really reasonably should *not* copy the attributes of the designated node, since it designates, through processing, the creation of a new context for that node's contents. The most compelling example of where you would seriously not want to automatically copy over all the attributes is the presence of a unique identifier. Would you want an `@xml:id` from the copied node to come in on the provided node during processing?

We found that a copied node is a new instantiation of its original, not identical to it. It took some time and careful review to clarify this. Every example we saw indicated that the notion of a copy was *not* identical to the origin. So we concluded that rather than change every example to match the behavior requested in this ticket that no, we needed clarification about the concept of the copy in TEI.

I do find problematic the name of the attribute since it most certainly does encourage confusion with the XSLT "copy-of". I wish we would rename this attribute, or provide a second attribute (`@copy`) to make room for processing the way you wanted on `@copyOf`. If we did that, we would need to provide a lot of explanation with caution about choosing a reasonable behavior and not bringing over unique identifiers. But Council did not agree that we would want two attributes and I think mainly did not want to encourage this approach because it would stand to add more confusion. Particularly it seemed people did not favor making TEI's `@copyOf` mirror XSLT's copy-of. I don't like the inevitable confusion of those two, but I do ultimately agree with Council's decision. We should not define this attribute to require the automatic porting of attributes from the source node. 

In cases where you do want the attributes, you can always introduce your own customization to introduce your own desired behavior in processing and document it in the header as you say. We just do not think that this should be the default expected behavior in processing.

**Comment by hcayless** [comment]
I would like to see the actual reasoning applied rather than a vague pointer to the Graphs chapter, in which I don't see any examples where the target of `@copyOf` has attributes other than `@xml:id`. If that's the extent of the argument, I'm afraid I'm not convinced. 4 is, as I pointed out, the worst option. I don't understand the analogy to XSLT's copy-of—it's a shame they have basically the same name, but it is what it is. Obviously, the element with `@copyOf` isn't the original. Perhaps there's a fundamental disagreement as to the nature of node content going on here, but my position is that attributes are content too. I could even make a case that an `@xml:id` on the target implies that the copying element ought to have one too, even if it can't be the same. 3 is probably the most rational option, if the most work. 2, and 1 are justifiable. I can't see any argument that the option that forces the encoder to do the most work is the best one. Literally, no such argument has been made, just a gesture towards examples that don't seem to apply, where target attributes can't be copied because there aren't any except `@xml:id`s.

I can make an argument for preserving target attributes in the case of transposition, where a line in one source has been moved in another. You would typically want attributes such as the line number / name of the line to stick around. See, e.g., https://www.tei-c.org/release/doc/tei-p5-doc/en/html/TC.html#index-body.1_div.13_div.4. I imagine you'll want to "fix" that, though it seems correct to me.

If I can list the options again:

1. Implementer's choice as to how attributes behave—potentially messy, but fine.
2. Attributes are copied, mutatis mutandis (ids have to be changed). So attributes behave like content, to the extent they can.
3. Attributes are merged. Ones declared on the pointing element win. This seems the most logical and gives encoders the most power, though it does open the question of whether this is a real copy if the attributes are different.
4. No attributes are to be copied, meaning encoders have to duplicate attributes themselves. Does at least provide clarity, but this feels like the least intuitive option as well as being the one that requires the most work on the part of the encoder. It also means attributes are treated opposite to content, which feels intuitively wrong.

So I'd love to see the argument for 4, which seems to me the most obviously wrong approach. To be absolutely clear, this is not functionality that Council needs to implement themselves. It has no bearing on Guidelines processing. It _might_ be something the Stylesheets could address and I suppose there might be new rules that could be introduced around this. I would encourage you to adopt a "first, do no harm" approach and leave it alone (i.e. pick number 1) rather than break existing work.

**Comment by martindholmes** [comment]
@hcayless  I checked all our mentions of @copyOf, and it talks ONLY of “content”. In the XML 1.1 spec:

`content ::= CharData? ((element | Reference | CDSect | PI | Comment) CharData?)*`

So attributes are definitely not included by definition. I think it's reasonable to ask for clarification on this, and I think we are clarifying it according to the original intent. If you think we should change this definition rather than clarify and reassert it, that would be a separate ticket, but I believe the Guidelines already did take a position on attributes, albeit perhaps by omission.

**Comment by sydb** [comment]
BTW, the same definition is in XML 1.0 5th ed. (Production #43.)

**Comment by hcayless** [comment]
@martindholmes That is at least an argument, thank you. I'd want to see some definitions in the Guidelines where it says the use of the word "content" is equivalent to the production in the XML grammar. Without such a definition, I'm afraid I don't accept that argument. The production name could be "foo" and the grammar would still work. But even if you're right, surely the correct approach is to take no position on it and leave it to implementers, not to say that any implementers who are copying attributes are doing it wrong. That seems like a major policy reversal for TEI. The original impetus for this issue was to gain clarity over what to do with attributes in this situation, so to say that another issue is required seems a bit rich.

**Comment by martindholmes** [comment]
@hcayless Point taken. I guess the question is whether devising the original intent of the creators of the attribute is a reasonable approach, or whether a perceived vagueness opens it up to reinterpretation. I think taking the position that it's up to implementers is slightly clearer than doing nothing, but it would make the attribute a barrier to interchange unless we then create a machine-processable mechanism for declaring whatever your own personal approach is.

Whenever we tighten up a vague point in the guidelines, we inevitably risk causing some end-users' implementations to be implicitly deprecated, but it is always possible to pin your ODD to a prior version of P5 which allows you the latitude you need.

As you can see on the ticket, I was originally in the camp that advocated copying attributes (excepting `@xml:id`), but re-reading all the Glines prose and specs changed my mind, because it seemed to me that most people reading closely would come to the same conclusion that attributes were not included. I don't think anyone on Council is strongly wedded to any particular position, but in the absence of any other evidence, I think assuming that "content" is used in the XML Spec manner is not unreasonable.

**Comment by joeytakeda** [comment]
@hcayless I'm in complete agreement and tried (but failed) to argue for approach 1 (and I was quite happy to see your comment, as Council can attest!). I was (and continue to be) a big advocate for the "leave it to implementers" approach precisely for the reasons you give: the TEI has been vague, and implementers may have copied attributes (as I have), or ignored attributes, or merged attributes for however long that this has been part of the spec (and indeed, the Stylesheets themselves don't do anything, which to me seems to imply that it is up for implementers).

I think this mechanism will need rethinking in P6, but for the time being, we should just accept that we cede that responsibility to implementers (and probably not build a new mechanism, but rather just recommend a remark saying that ones use of copyOf should be documented in their ODD)

**Comment by hcayless** [comment]
@martindholmes I do see your point, but a lack of mention doesn't equate to a prohibition. It just means they hadn't thought about it that hard. This is one of those areas that was under-specified for a long time and if it can't be tightened up in a satisfactory way, then it's better to leave it alone. The graph stuff in my mind never made it out of the half-baked stage—it's just never had enough of a workout from the community to become mature, so I wouldn't be doing exegesis on it. @joeytakeda I'm very relieved that it's not just me! 🙏

**Comment by lb42** [comment]
No, Hugh, you are not alone.

**Comment by martindholmes** [comment]
@lb42 Do you remember the original intent?

**Comment by lb42** [comment]
All of those  att.linking attributes were proposed and defined by terry langendoens ai workgroup (in those days ai was short  for "analysis and interpretation "(rather thanspawn of thedevil)). Check working papers of the group.

As regards the uselessness or otherwise of the graphs and trees module i note only that when it was last proposed to remove th
emodule  on the grounds that nobody used it, several users popped up saying that actually they did so use it.

**Comment by ebeshero** [comment]
@lb42 Thanks for this helpful info. We'll take a look!
